server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -322,7 +325,13 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
   {
     try {
       log.info("Loading segment %s", segment.getIdentifier());
-      if (loadSegment(segment, callback)) {
+      if (segmentsToDelete.contains(segment)) {
+        synchronized (lock) {
+          segmentsToDelete.remove(segment);
+        }
+      }
+      loadSegment(segment, callback);
```

Comment:
would be nice to add a javadoc for loadSegment method that the return value should not be used, or even better to modify the loadSegment to return void and add javadoc to isAnnounced() to make sure we dont use its returned value elsewhere.